File: data-engineering/nyc-taxi-workshop/code/notebooks/01-General/1-MountBlobStorage.scala
// Databricks notebook source
// MAGIC %md
// MAGIC # What's in this exercise?
// MAGIC 
// MAGIC Learn how to mount various blob storage container directories so you can access them easily in your code, without defining credentials and wasbs URIs in each notebook

// COMMAND ----------

// MAGIC %md
// MAGIC ### 1. Define credentials
// MAGIC We defined storage account credentials at the cluster level to work with the database UI of Azure databricks.<BR>
// MAGIC Here we need it to mount blob storage

// COMMAND ----------

val storageAccount = "<yourStorageAccountName>"
val storageAccountKey= spark.conf.get("spark.hadoop.fs.azure.account.key." + storageAccount + ".blob.core.windows.net")

// COMMAND ----------

// MAGIC %md
// MAGIC ### 2. Mount blob storage

// COMMAND ----------

def mountStorageContainer(storageAccount: String, storageAccountKey: String, storageContainer: String, blobMountPoint: String)
{
   try {
     
     println(s"Mounting ${storageContainer} to ${blobMountPoint}:")
    // Unmount the storage container if already mounted
    dbutils.fs.unmount(blobMountPoint)

  } catch { 
    //If this errors, the container is not mounted
    case e: Throwable => println(s"....Container is not mounted; Attempting mounting now..")

  } finally {
    // Mount the storage container
    val mountStatus = dbutils.fs.mount(
    source = "wasbs://" + storageContainer + "@" + storageAccount + ".blob.core.windows.net/",
    mountPoint = blobMountPoint,
    extraConfigs = Map("fs.azure.account.key." + storageAccount + ".blob.core.windows.net" -> storageAccountKey))
  
    println("...Status of mount is: " + mountStatus)
  }
}

// COMMAND ----------

mountStorageContainer(storageAccount,storageAccountKey,"nyctaxi-staging","/mnt/data/nyctaxi/stagingDir")
mountStorageContainer(storageAccount,storageAccountKey,"nyctaxi-ref-data","/mnt/data/nyctaxi/referenceDataDir")
mountStorageContainer(storageAccount,storageAccountKey,"nyctaxi-raw","/mnt/data/nyctaxi/rawDir")
mountStorageContainer(storageAccount,storageAccountKey,"nyctaxi-curated","/mnt/data/nyctaxi/curatedDir")
mountStorageContainer(storageAccount,storageAccountKey,"nyctaxi-consumption","/mnt/data/nyctaxi/consumptionDir")

// COMMAND ----------

// MAGIC %md
// MAGIC ### 3. List storage containers mounted

// COMMAND ----------

display(dbutils.fs.ls("/mnt/data/nyctaxi/"))

// COMMAND ----------

// MAGIC %md
// MAGIC ### 4. Unmount blob storage - how to

// COMMAND ----------

//Unmount for notebook execution 
/*
dbutils.fs.unmount("/mnt/data/nyctaxi/stagingDir")
dbutils.fs.unmount("/mnt/data/nyctaxi/refDataDir")
dbutils.fs.unmount("/mnt/data/nyctaxi/rawDir")
dbutils.fs.unmount("/mnt/data/nyctaxi/curatedDir")
dbutils.fs.unmount("/mnt/data/nyctaxi/consumptionDir")

*/

// COMMAND ----------

//Refresh mounts
dbutils.fs.refreshMounts()

// COMMAND ----------

//List
display(dbutils.fs.ls("/mnt/data/nyctaxi/"))

// COMMAND ----------

// MAGIC %md
// MAGIC ### 4. Refresh mount points

// COMMAND ----------

//Refresh mounts
dbutils.fs.refreshMounts()

// COMMAND ----------

// MAGIC %fs
// MAGIC ls /mnt/data/nyctaxi/

// COMMAND ----------

File: data-engineering/nyc-taxi-workshop/code/notebooks-py/00-HowTo/1-MountBlobStorage.py
# Databricks notebook source
# MAGIC %md
# MAGIC # Mount blob storage
# MAGIC 
# MAGIC Mounting blob storage containers in Azure Databricks allows you to access blob storage containers like they are directories.<BR>

# COMMAND ----------

# MAGIC %md
# MAGIC ### 1. Define credentials
# MAGIC To mount blob storage - we need storage credentials - storage account name and storage account key

# COMMAND ----------

storageAccount = "<<your_storage_account>>"
storageAccountConfKey = "fs.azure.account.key." + storageAccount + ".blob.core.windows.net"
storageAccountConfVal= spark.conf.get("spark.hadoop.fs.azure.account.key." + storageAccount + ".blob.core.windows.net")

# COMMAND ----------

# MAGIC %md
# MAGIC ### 2. Mount blob storage

# COMMAND ----------

dbutils.fs.mount(
  source = "wasbs://nyctaxi-scratch@{}.blob.core.windows.net/".format(storageAccount),
  mount_point = "/mnt/<userid>/data/nyctaxi/scratchDir",
  extra_configs = {storageAccountConfKey: storageAccountConfVal})

# COMMAND ----------

# Display directories
display(dbutils.fs.ls("/mnt/<userid>/data/nyctaxi/"))

# COMMAND ----------

# MAGIC %md
# MAGIC ### 3. Refresh mount points

# COMMAND ----------

# Refresh mounts if applicable
dbutils.fs.refreshMounts()

# COMMAND ----------

# MAGIC %md
# MAGIC ### 4. How to unmount

# COMMAND ----------

dbutils.fs.unmount("<yourMountPoint>")